src/fmeval/transforms/semantic_perturbations.py

```diff
@@ -0,0 +1,301 @@
+    QUERTY_KEY_APPROX["y"] = "ytugjhrfvbnji"
+    QUERTY_KEY_APPROX["u"] = "uyihkjtgbnmlo"
+    QUERTY_KEY_APPROX["i"] = "iuojlkyhnmlp"
+    QUERTY_KEY_APPROX["o"] = "oipklujm"
+    QUERTY_KEY_APPROX["p"] = "plo['ik"
+
+    QUERTY_KEY_APPROX["a"] = "aqszwxwdce"
+    QUERTY_KEY_APPROX["s"] = "swxadrfv"
+    QUERTY_KEY_APPROX["d"] = "decsfaqgbv"
+    QUERTY_KEY_APPROX["f"] = "fdgrvwsxyhn"
+    QUERTY_KEY_APPROX["g"] = "gtbfhedcyjn"
+    QUERTY_KEY_APPROX["h"] = "hyngjfrvkim"
+    QUERTY_KEY_APPROX["j"] = "jhknugtblom"
+    QUERTY_KEY_APPROX["k"] = "kjlinyhn"
+    QUERTY_KEY_APPROX["l"] = "lokmpujn"
+
+    QUERTY_KEY_APPROX["z"] = "zaxsvde"
+    QUERTY_KEY_APPROX["x"] = "xzcsdbvfrewq"
+    QUERTY_KEY_APPROX["c"] = "cxvdfzswergb"
+    QUERTY_KEY_APPROX["v"] = "vcfbgxdertyn"
+    QUERTY_KEY_APPROX["b"] = "bvnghcftyun"
+    QUERTY_KEY_APPROX["n"] = "nbmhjvgtuik"
+    QUERTY_KEY_APPROX["m"] = "mnkjloik"
+    QUERTY_KEY_APPROX[" "] = " "
+
+    def __init__(
+        self,
+        input_keys: List[str],
+        output_keys: List[str],
+        num_perturbations: int = 5,
+        seed: int = 5,
+        perturbation_prob: float = 0.1,
+    ):
+        """ButterFinger initializer.
+
+        :param input_keys: A single-element list containing the key corresponding to the text input to be perturbed.
+        :param output_keys: The keys corresponding to perturbed text outputs generated by the `perturb` method.
+        :param num_perturbations: The number of perturbed outputs to generate via the `perturb` method.
+            Note that the number of output keys must match this parameter.
+        :param seed: A random seed, used by pseudorandom number generators.
+        :param perturbation_prob: The probability that a given character in the input text will be perturbed.
+        """
+        super().__init__(
+            input_keys,
+            output_keys,
+            num_perturbations,
+            seed,
+            perturbation_prob=perturbation_prob,
+        )
+        self.perturbation_prob = perturbation_prob
+
+    def perturb(self, text: str) -> List[str]:
+        """Return a list where each element is a copy of the original text after undergoing perturbations.
+
+        The perturbations mimic keyboard-induced typos and are applied (with a certain probability)
+        to each character in `text`.
+
+        :param text: The input text to be perturbed.
+        :returns: A list of perturbed text outputs.
+        """
+        prob_of_typo = int(self.perturbation_prob * 100)
+        perturbed_texts = []
+        for _ in range(self.num_perturbations):
+            butter_finger_text = []
+            for letter in text:
+                lowercase_letter = letter.lower()
+                if lowercase_letter not in ButterFinger.QUERTY_KEY_APPROX.keys():
+                    new_letter = lowercase_letter
+                else:
+                    if random.choice(range(0, 100)) <= prob_of_typo:
```

Comment:
Question: Why do we use `random.choice` here to pick a random whole number? Why not use `random.random()`? 

There's an edge case when a user puts 0.99 or larger as the probability that will always result in a typo because the largest number from the range is 99, making the condition always true. Using `random.random()` will cover this edge case and be closer to the probability the user requests without needing to round. 

There's also another edge case if the user sets probability to 0.00, there is a 1% chance of a typo, since there's a 1/100th chance of picking 0 randomly. 

To cover both cases, the condition should be replaced with `random.random() < prob_of_typo`.